foolbox/batch_attacks/iterative_projected_gradient.py

```diff
@@ -422,6 +457,86 @@ def as_generator(self, a,
 L1BasicIterativeAttack.__call__.__doc__ = L1BasicIterativeAttack.as_generator.__doc__
 
 
+class SparseL1BasicIterativeAttack(
+        SparseL1GradientMixin,
+        L1ClippingMixin,
+        L1DistanceCheckMixin,
+        GDOptimizerMixin,
+        IterativeProjectedGradientBaseAttack):
+
+    """Sparse version of the Basic Iterative Method
+    that minimizes the L1 distance introduced in [1]_.
+
+     References
+    ----------
+    .. [1] Florian Tramèr, Dan Boneh,
+           "Adversarial Training and Robustness for Multiple Perturbations",
+           https://arxiv.org/abs/1904.13000
+
+    .. seealso:: :class:`L1BasicIterativeAttack`
+
+    """
+
+    @generator_decorator
+    def as_generator(self, a,
+                     q=80.0,
+                     binary_search=True,
+                     epsilon=0.3,
+                     stepsize=0.05,
+                     iterations=10,
+                     random_start=False,
+                     return_early=True):
+
+        """Sparse version of a gradient-based attack that minimizes the
+        L1 distance.
+
+        Parameters
+        ----------
+        inputs : `numpy.ndarray`
+            Batch of inputs with shape as expected by the underlying model.
+        labels : `numpy.ndarray`
+            Class labels of the inputs as a vector of integers in [0, number of classes).
+        unpack : bool
+            If true, returns the adversarial inputs as an array, otherwise returns Adversarial objects.
+        q : float
+            Relative percentile to make gradients sparse (must be in [0, 100))
+        binary_search : bool or int
+            Whether to perform a binary search over epsilon and stepsize,
+            keeping their ratio constant and using their values to start
+            the search. If False, hyperparameters are not optimized.
+            Can also be an integer, specifying the number of binary
+            search steps (default 20).
+        epsilon : float
+            Limit on the perturbation size; if binary_search is True,
+            this value is only for initialization and automatically
+            adapted.
+        stepsize : float
+            Step size for gradient descent; if binary_search is True,
+            this value is only for initialization and automatically
+            adapted.
+        iterations : int
+            Number of iterations for each gradient descent run.
+        random_start : bool
+            Start the attack from a random point rather than from the
+            original input.
+        return_early : bool
+            Whether an individual gradient descent run should stop as
+            soon as an adversarial is found.
+        """
+
+        assert epsilon > 0
+
+        assert 1.0 <= q <= 99.0, '`q` must be in [0, 100).'
```

Comment:
How about `< 100` instead of `<= 99.0` ?

Reply:
And actually `== 100` should also be fine (from NumPy's side)

Reply:
If one uses `q==100`, then all entries of the gradient will be set to zero because the mask calculated in  [line 213](https://github.com/bethgelab/foolbox/pull/348/files/d0d5cfa178a0bb98620b09a19196b1f4eb9d2662#diff-ad909d83c970d224f1054cb941eb3fd2R213) contains all pixels, won't it?

Reply:
True, it's that way round. So please change `<= 99` to `< 100`.

And `>= 1` should be `>= 0`, or not? Wouldn't it be fine to not mask any of them?